services/src/pro/datasets/in_memory.rs

```diff
@@ -0,0 +1,383 @@
+        StaticMetaData<MockDatasetDataSourceLoadingInfo, VectorResultDescriptor>,
+    >,
+    gdal_datasets:
+        HashMap<InternalDatasetId, Box<dyn MetaData<GdalLoadingInfo, RasterResultDescriptor>>>,
+    uploads: HashMap<UploadId, Upload>,
+}
+
+impl DatasetDb<UserSession> for ProHashMapDatasetDb {}
+
+#[async_trait]
+impl DatasetProviderDb<UserSession> for ProHashMapDatasetDb {
+    async fn add_dataset_provider(
+        &mut self,
+        _session: &UserSession,
+        _provider: Box<dyn DatasetProviderDefinition>,
+    ) -> Result<DatasetProviderId> {
+        todo!()
+    }
+
+    async fn list_dataset_providers(
+        &self,
+        _session: &UserSession,
+        _options: Validated<DatasetProviderListOptions>,
+    ) -> Result<Vec<DatasetProviderListing>> {
+        todo!()
+    }
+
+    async fn dataset_provider(
+        &self,
+        _session: &UserSession,
+        _provider: DatasetProviderId,
+    ) -> Result<Box<dyn DatasetProvider>> {
+        todo!()
+    }
+}
+
+pub trait ProHashMapStorable: Send + Sync {
+    fn store(&self, id: InternalDatasetId, db: &mut ProHashMapDatasetDb) -> TypedResultDescriptor;
+}
+
+impl DatasetStorer for ProHashMapDatasetDb {
+    type StorageType = Box<dyn ProHashMapStorable>;
+}
+
+impl ProHashMapStorable for MetaDataDefinition {
+    fn store(&self, id: InternalDatasetId, db: &mut ProHashMapDatasetDb) -> TypedResultDescriptor {
+        match self {
+            MetaDataDefinition::MockMetaData(d) => d.store(id, db),
+            MetaDataDefinition::OgrMetaData(d) => d.store(id, db),
+            MetaDataDefinition::GdalMetaDataRegular(d) => d.store(id, db),
+            MetaDataDefinition::GdalStatic(d) => d.store(id, db),
+        }
+    }
+}
+
+impl ProHashMapStorable for StaticMetaData<OgrSourceDataset, VectorResultDescriptor> {
+    fn store(&self, id: InternalDatasetId, db: &mut ProHashMapDatasetDb) -> TypedResultDescriptor {
+        db.ogr_datasets.insert(id, self.clone());
+        self.result_descriptor.clone().into()
+    }
+}
+
+impl ProHashMapStorable
+    for StaticMetaData<MockDatasetDataSourceLoadingInfo, VectorResultDescriptor>
+{
+    fn store(&self, id: InternalDatasetId, db: &mut ProHashMapDatasetDb) -> TypedResultDescriptor {
+        db.mock_datasets.insert(id, self.clone());
+        self.result_descriptor.clone().into()
+    }
+}
+
+impl ProHashMapStorable for GdalMetaDataRegular {
+    fn store(&self, id: InternalDatasetId, db: &mut ProHashMapDatasetDb) -> TypedResultDescriptor {
+        db.gdal_datasets.insert(id, Box::new(self.clone()));
+        self.result_descriptor.clone().into()
+    }
+}
+
+impl ProHashMapStorable for GdalMetaDataStatic {
+    fn store(&self, id: InternalDatasetId, db: &mut ProHashMapDatasetDb) -> TypedResultDescriptor {
+        db.gdal_datasets.insert(id, Box::new(self.clone()));
+        self.result_descriptor.clone().into()
+    }
+}
+
+#[async_trait]
+impl DatasetStore<UserSession> for ProHashMapDatasetDb {
```

Comment:
`typedef ProDatasetStore = DatasetStore<UserSession>`? also for other DBS?

Reply:
#later